docs/development-tips.md

```diff
@@ -0,0 +1,254 @@
+
+It's often useful to be able to inspect what the output of an intermediate LLB stage would look like after it has executed, and you can perform an analog to `printf` debugging by solving the LLB stage to a Docker image and then inspecting the resulting image:
+
+```go
+// DEBUG: Solve the LLB stage to a Docker image.
+if err := buildkit.SolveToDocker(ctx, dm.config.Client, &<llb.Stage>, dm.config.ConfigData, dm.config.ImageName+"-<llb.Stage suffix>"); err != nil {
+    return nil, err
+}
+```
+
+For example, if you want to see what the resulting Docker image looks like at the `busyBoxApplied` stage, you can add the `buildkit.SolveToDocker` call to the end of the `busyBoxApplied` stage as follows. The result will be a Docker image with the `-busyBoxApplied` suffix to the tag that you can inspect with the `docker` CLI or `dive` tool:
+
+```go
+busyBoxApplied := dm.config.ImageState.File(llb.Copy(toolImage, "/bin/busybox", "/bin/busybox"))
+if err := buildkit.SolveToDocker(ctx, dm.config.Client, &busyBoxApplied, dm.config.ConfigData, dm.config.ImageName+"-busyBoxApplied"); err != nil {
+    return nil, err
+}
+```
+
+## Inspect a Docker image
+
+### Use `docker` to inspect the metadata of the image
+
+For a quick check of a Docker image, the built in `docker` CLI commands can be useful:
+
+- [`docker inspect`](https://docs.docker.com/engine/reference/commandline/inspect/) can show you the metadata for the image and verifying that the patching process generally preserves the metadata of the original image.
+- [`docker history`](https://docs.docker.com/engine/reference/commandline/history/) can give you a quick overview of the layers in the image, and with the `--no-trunc` flag, can provide the commands that were run to create each layer.
+
+### Use `dive` to inspect the filesystem differences at each layer of the image
+
+Instructions for installing and using the `dive` CLI tool are at https://github.com/wagoodman/dive.
+
+`dive` provides a simple interface for walking the layers of the image and inspecting the files that were added or changed at each layer with your arrow keys.
+
+- `Tab` will toggle between navigating the layers and the files in the layer.
+- Filtering out unmodified files with `Ctrl+U` while in files view will effectively show you the file diff introduced by that layer.
+
+In particular, if you are adding or changing any of the patching functionality, the diff view of the files in the image can be useful to verify that the expected files have actually been written to the target image.
+
+### Extract files from the image to inspect them
+
+`dive` won't let you read the contents of the files in the image though; to do that, you can use the `docker cp` command to copy the files out of the image to a local folder. Note that `docker cp` only works with containers and not just container images, so you will need to create a container from the image and then copy the files out of the container:
```

Comment:
I have been using `crane export` to dump the filesystem. Should we mention this instead of `docker cp`?

You can also see diffs using
```
 diff \ 
    <(crane export foo:bar - | tar -tvf - | sort) \
    <(crane export foo:bar-patched - | tar -tvf - | sort)
 ```

Reply:
I'll add it to the doc because `crane export` is a good way to explore the entire filesystem interactively, but I switched away from extracting the entire image in my workflow because there's usually only a couple of files I want to inspect, and the `dive` exploration view is really handy for the per layer diffing.

Also, neat `diff` command line, thanks for sharing that!